libbeat/ebpf/watcher_linux.go

```diff
@@ -0,0 +1,183 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build linux
+
+package ebpf
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/elastic/ebpfevents"
+)
+
+var (
+	gWatcherOnce sync.Once
+	gWatcherErr  error
+	gWatcher     Watcher
+)
+
+type client struct {
+	name    string
+	mask    EventMask
+	records chan ebpfevents.Record
+}
+
+// EventMask is a mask of epbevents.EventType which is used to control which event types clients will receive.
+type EventMask uint64
+
+// Watcher observes kernel events, using ebpf probes from the ebpfevents library, and sends the
+// events to subscribing clients.
+//
+// A single global watcher can exist, and can deliver events to multiple clients. Clients subscribe
+// to the watcher, and all ebpf events that match their mask will be sent to their channel.
+type Watcher struct {
+	sync.Mutex
+	cancel  context.CancelFunc
+	loader  *ebpfevents.Loader
+	clients map[string]client
+	status  status
+}
+
+type status int
+
+const (
+	stopped status = iota
+	started
+)
+
+// GetWatcher creates the watcher, if required, and returns a reference to the global Watcher.
+func GetWatcher() (*Watcher, error) {
+	gWatcher.Lock()
+	defer gWatcher.Unlock()
+
+	// Try to load the probe once on startup so consumers can error out.
+	gWatcherOnce.Do(func() {
+		if gWatcher.status == stopped {
+			l, err := ebpfevents.NewLoader()
+			if err != nil {
+				gWatcherErr = fmt.Errorf("init ebpf loader: %w", err)
```

Comment:
Is it necessary to use a global for this? If it needs to be retained between calls (though why? I don't see it being accessed anywhere) it can be kept in the `Watcher`.

Reply:
I've added these to Watcher

Reply:
I've moved `err` into the `Watcher` struct.

This is needed so that subsequent callers of `GetWatcher` can get the same error as the first call. The most likely cause of the error is that the system doesn't support ebpf, so there's no point retrying, but the other callers should still get the error.  So the error needs to be kept in a global. But if you know a better way to do this, I can change

Reply:
Yeah, OK. Keeping it in the struct is the conventional approach to make errors sticky. Then you can check the err field on method entry and return the error if it is non-nil. You can also have an `Err() error` method to return the error if you think it will be helpful (probably not here).

Because of the prevalence of concurrency in Go, use of globals is avoided, and if it can't be, try to make any singleton have only one entry (having the error as an associated global essentially makes the singleton doubly rooted).